### DISCUSSION TITLE
Trying to use puppeteer in vercel deployment

### DISCUSSION THREAD

**Discussion by abhishekmg** [vercel/next.js#63961] [Help]
### Summary

I am trying to use puppeteer to launch a chrome browser in next js api , but this is not working when I deploy to Vercel. This is working in my local machine which is a mac using Apple M1 and macos monterey.

The vercel node version is 20

The code can be found in this repo https://github.com/abhishekmg/screenshot-test

The screenshot logic is in src/api/route.js

Iam getting this error when I push to vercel

![image](https://github.com/vercel/next.js/assets/34393560/e754fde6-4dca-4813-b41f-be3af994c709)

```
import { launch } from "puppeteer";

const takeScreenshot = async (url) => {
  // this works on my local machine
  const browser = await launch();

  const page = await browser.newPage();

  await page.goto(url, { waitUntil: "load" });

  const buffer = await page.screenshot();

  await page.close();
  await browser.close();

  return buffer;
};

export async function GET(request) {
  const buffer = await takeScreenshot(
    "https://www.npmjs.com/package/puppeteer"
  );

  return Response.json({ data: buffer });
}
```

I tried using `chrome-aws-lamda` as mentioned https://github.com/orgs/vercel/discussions/124, but still facing issues

### Example

https://github.com/abhishekmg/screenshot-test

**Comment by nicnocquee** [comment]
I tried this last year and it worked https://github.com/nicnocquee/instagram-puppet/tree/main but need to use the specific version of the chrome-aws-lambda and puppeteer-core

**Reply by abhishekmg** [reply]
I tried with the versions of the chrome-aws-lambda and puppeteer-core mentioned in the repo but I still get the same error 😭

**Reply by nicnocquee** [reply]
sorry bout that, it worked last year 😅

**Reply by abhishekmg** [reply]
no problem ! it might be because Vercel linux machines this past year , but to exactly sure

**Comment by estyxx** [comment]
Hey, did you find any solution?

**Comment by galbahar** [comment]
any solution?

**Comment by subhadeeproy3902** [comment]
any solution ?

**Comment by vieiralucas** [comment]
This is a recurring issue — Puppeteer works locally because you have Chrome installed, but Vercel's serverless environment doesn't include it. The `@sparticuz/chromium` workaround helps but is fragile (breaks across Node/Next.js version bumps, hits bundle size limits).

If you're primarily generating PDFs rather than doing browser automation, there are alternatives that skip Chromium entirely:

- **[Gotenberg](https://gotenberg.dev)** — open source, self-hosted in Docker. Full rendering quality, you manage the container.
- **[PDFShift](https://pdfshift.io)** — hosted API, free tier.
- **[Presa](https://presa.dev)** — hosted API I built, free tier (100 PDFs/month).

You send HTML, get a PDF back — no Chromium binary needed in your Vercel function:

```typescript
import { PDF } from "presa-sdk";
const { buffer } = await PDF.generate({ html, format: "A4" });
```

Deploys to Vercel without issues since it's just an HTTP call.

**Comment by JEETJM** [comment]
When deploying Next.js with Puppeteer on Vercel, the issue usually happens because Vercel serverless functions do not include a full Chromium environment.

A common solution is to use `puppeteer-core` together with `chrome-aws-lambda`, which works better in serverless environments.

Steps you can try:

1. Install required packages:
   npm install puppeteer-core chrome-aws-lambda

2. Update your code to launch Chromium correctly in a serverless environment.

Example:

```javascript
import chromium from "chrome-aws-lambda";
import puppeteer from "puppeteer-core";

export default async function handler(req, res) {
  const browser = await puppeteer.launch({
    args: chromium.args,
    executablePath: await chromium.executablePath,
    headless: chromium.headless
  });

  const page = await browser.newPage();
  await page.goto("https://example.com");

  const title = await page.title();

  await browser.close();

  res.status(200).json({ title });
}

**Comment by icyJoseph** [comment]
There's a guide here:

- https://vercel.com/kb/guide/deploying-puppeteer-with-nextjs-on-vercel
- Source code: https://github.com/gabenunez/puppeteer-on-vercel
